rllib/examples/rlhf/ppo_ft/rlhf_env.py

```diff
@@ -0,0 +1,206 @@
+from typing import Any
+import gymnasium as gym
+from ray.rllib.env.env_context import EnvContext
+
+import transformers
+import datasets
+import numpy as np
+import torch
+from ray.rllib.utils.spaces.repeated import Repeated
+import gymnasium.spaces as sp
+
+
+def generate_response(
+    model: torch.nn.Module, 
+    *, 
+    input_ids: torch.tensor, 
+    max_length:int, 
+    eos_token_id: int
+):
+    """Generate a response using the model."""
+    generated_sequence = []
+    probs_list = []
+    model_in = torch.clone(input_ids)
+    with torch.no_grad():
+        for i in range(max_length):
+            # Get the logits using the forward() method
+            logits = model(model_in).logits
+
+            # Get the logits of the last token
+            next_token_logits = logits[:, -1, :]
+
+            # Apply a softmax function to the logits to get the probabilities
+            probs = next_token_logits.softmax(dim=-1)
+
+            # Sample the next token from the probability distribution
+            next_token = torch.multinomial(probs, num_samples=1)
+
+            # Append the probabilities and the generated token
+            generated_sequence.append(next_token)
+            probs_list.append(probs)
+
+            # Update the input_ids with the generated token
+            model_in = torch.cat([model_in, next_token], dim=-1)
+
+            if next_token.item() == eos_token_id:
+                break
+
+        # Decode and print the generated sequence
+        generated_tokens = torch.cat(generated_sequence, dim=-1)
```

Comment:
if we are doing sequences in 1 shot, how are we gonna take discounted future return into consideration?
should the reward function score every single token generated?

Reply:
I guess to clarify my confusion, for example, the kind of behavior we are trying to encourage is when we are in the middle of a sentence, the chosen token should allow us to conclude the whole sentence sooner.
so the training sample should only be the partial sentence, and the next word choice?
I don't quite get how scoring the whole sequence would help.

Reply:
This is how instructGPT introduces RLHF. the truth is that you don't get any evaluation signal until the end of the sentence. i.e. the reward model is queried only when the answer is complete to evaluate whether the response was good for the input instruction. So your discount factor should be 1 here because you want to propagate that final reward equally to every token decision you have made so far. Also this way you'll only do one forward pass on RM and b/c you don't need to call it auto-regressively. However if it's a conversation things differ, the model can ask for clarifications etc. but in conversation you also have to generate full sentences on each turn. If you incorporate RLHF in conversation fashion, then your model can learn to ask you clarifying questions, because that will help it give more satisfying answers faster.

Reply:
oh. I see.